### DISCUSSION TITLE
Proposed RFC Suggestion: Expose SaveData Gem to Script Canvas

### DISCUSSION THREAD

**Discussion by ppinfel** [o3de/sig-content#33] [Ideas]
Currently, Script Canvas has no way to save and load data which is a common feature in games, such as saving game settings that should persist between game sessions. 

The SaveData Gem already implements this but is only available in C++. In order for Script Canvas to have access to it, the SaveData Gem interfaces(?) should be exposed to Script Canvas.

**Comment by nick-l-o3de** [comment]
Ah, to get this in, the SaveData gem would have to be upgraded to reflect not just to the serialize and edit context, but to the Behavior Context as well (Which is what Script Canvas and LUA reads from).

**Comment by bosnichd** [comment]
So the SaveData Gem interface was intentionally designed to be limited to "save a raw data buffer", "load a raw data buffer" (with extra bits to give each buffer a unique name, potentially save for a specific user, get a callback/notification when it's done saving or loading, and some convenience functions that let you save/load any object that has been exposed to the Serialize Context), because we don't want the low level save/load API to prescribe any specific data format.

Having said that, the ability to save/load things from Script Canvas would obviously be very useful, and we can create some custom Script Canvas nodes to do so, but the question is how would the user get the data they want saved/loaded into/out of the nodes?

I spoke with @lsemp3d about it, who proposed a general purpose way to have a JSON file as a data schema that can be referenced in ScriptCanvas as any type of data payload, which would make it fairly straightforward to add custom Save/Load SC nodes to pass in one of these objects and have it marshal the data into a buffer.
